### PR TITLE
Merged binary

### PR DESCRIPTION
**Pull request by apfohl**
This PR unifies the three binaries `ecdsakeygen`, `ecdsasign` and `ecdsaverify` into a single binary `ecdsautils`.

### CODE REVIEW

**Comment by tcatm** [comment]
Please describe your changes in more detail in the commit message. Without reviewing the changes I can not tell whether the README needs to be updated, too.

**Comment by apfohl** [comment]
Ok, I will do by the end of the week.

**Comment by apfohl** [comment]
All right, I wrote an explaining commit message and fixed the readme according to the new behavior.

**Comment by tcatm** [comment]
Thanks. It's much clearer now.

As we're changing the syntax anyway I'd like to split `keygen` into `generate-key` and `show-key` and drop the `-s` and `-p` options so that there are distinct commands for both operations.

**Comment by apfohl** [comment]
Yes, good idea.

**Comment by apfohl** [comment]
Done. :beer:

**Comment by tcatm** [comment]
Great. I've had a look at the code and I think we should move `sign` and `verify` to their own files. `main()` is a little hard to read now.

I'm also pondering whether we should rename the binary to `ecdsautil` (dropping the s).

**Comment by apfohl** [comment]
I agree. Will implement that.

**Comment by apfohl** [comment]
Done again. I renamed it to `ecdsautil` and encapsulated all subcommands. :beers:

**Comment by tcatm** [comment]
This looks quite good already. What do you think about calling `sign(argc - 1, argv + 1)` (same for verify) instead of adjusting the pointers in `sign()`/`verify()`?

**Comment by apfohl** [comment]
Allright, anything else?

**Comment by tcatm** [comment]
Now you've moved the usage message for `sign` back into `main()`. I'd prefer `main()` to just pass any remaining arguments along.

**Comment by apfohl** [comment]
I think, that usage display and command line parsing is a job for ecdsautil.c, because it's the main program's task. I would go even further and strip every thing from `sign`, `verify`, `keygen` that has nothing explicitly to do with `sign`, `verify`, `keygen`. This would also be nice if one wants to implement a test suite for `ecdsautil` later (hinthint :)).

**Comment by tcatm** [comment]
We could do that abstraction in `sign()`. i.e. have `cmd_sign(argc, argv)` and `sign(filename)`.

**Comment by apfohl** [comment]
Ok, I will put it back to `sign` so that we can close this thread and come to optimization later. What do you think?

File: src/sign.c [hunk 1]
```diff
@@ -40,7 +41,7 @@
  * a uniformly random string, with the exception of the five bits
  * ecc_25519_gf_sanitize_secret sets and clears)
  */
-static void generate_k(uint8_t *k, const uint8_t prk[32], const uint8_t info[32]) {
```
Comment (tcatm) [inline src/sign.c]: Why did you remove `static` here?

Comment (apfohl) [inline src/sign.c]: Oh, I had to because I called the function from another file. I will fix it with the next commit.

File: src/ecdsautil.c [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  if (argc < 2) {
+    help();
+    return 1;
+  }
+
+  if (strcmp(argv[1], "help") == 0) {
+    help();
+    return 0;
+  } else if (strcmp(argv[1], "generate-key") == 0) {
+    new_secret();
```
Comment (tcatm) [inline src/ecdsautil.c]: I think we should call this `generate_key()` (and `show_key()`) to match the command name.

**Comment by apfohl** [comment]
Fixed.

File: src/ecdsautil.c [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    help();
+    return 0;
+  } else if (strcmp(argv[1], "generate-key") == 0) {
+    generate_key();
+    return 0;
+  } else if (strcmp(argv[1], "show-key") == 0) {
+    show_key();
+    return 0;
+  } else if (strcmp(argv[1], "sign") == 0) {
+    sign(argc, argv);
```
Comment (tcatm) [inline src/ecdsautil.c]: This should be `sign(argc - 1, argv + 1)`, too.

Comment (tcatm) [inline src/ecdsautil.c]: It should also return the return code of `sign()` so it matches the same pattern as `verify()`.

**Comment by tcatm** [comment]
I've noticed this removes reading the file to be verified from STDIN. Could you fix it to retain the old behaviour?

File: src/ecdsautil.c (hunk 3 above)
Comment (apfohl) [inline src/ecdsautil.c]: In order to do that, I need to change the error handling behavior. Right now in `error.h` calls `exit()`, which prevents returning a value from the sign function. In my opinion there's an inconsistency in the error handling.

**Comment by apfohl** [comment]
Sure.

**Comment by apfohl** [comment]
Ok, I recovered the original behavior.

**Comment by tcatm** [comment]
Great. How can I view the usage message of `verify`?

**Comment by apfohl** [comment]
As you did in the original `ecdsaverify`.

**Comment by apfohl** [comment]
Oh, I see there's a problem. Only if I type `ecsdautil verify --help` I get the usage message, because it exceeds the number of allowed parameters. That's a problem. It also was present in the original version `ecdsaverify`. Hmm, we shall improve. Any ideas how?
